plugins/outputs/postgresql/table_manager.go

```diff
@@ -0,0 +1,388 @@
+		tbl = &tableState{name: name}
+		tm.tables[name] = tbl
+	}
+	tm.tablesMutex.Unlock()
+	return tbl
+}
+
+func (tm *TableManager) refreshTableStructure(ctx context.Context, db dbh, tbl *tableState) error {
+	rows, err := db.Query(ctx, refreshTableStructureStatement, tm.Schema, tbl.name)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
+
+	cols := make(map[string]utils.Column)
+	for rows.Next() {
+		var colName, colTypeStr string
+		desc := new(string)
+		err := rows.Scan(&colName, &colTypeStr, &desc)
+		if err != nil {
+			return err
+		}
+
+		role := utils.FieldColType
+		switch colName {
+		case TimeColumnName:
+			role = utils.TimeColType
+		case TagIDColumnName:
+			role = utils.TagsIDColType
+		case TagsJSONColumnName:
+			role = utils.TagColType
+		case FieldsJSONColumnName:
+			role = utils.FieldColType
+		default:
+			// We don't want to monopolize the column comment (preventing user from storing other information there), so just look at the first word
+			if desc != nil {
+				descWords := strings.Split(*desc, " ")
+				if descWords[0] == "tag" {
+					role = utils.TagColType
+				}
+			}
+		}
+
+		cols[colName] = utils.Column{
+			Name: colName,
+			Type: utils.PgDataType(colTypeStr),
+			Role: role,
+		}
+	}
+	if err := rows.Err(); err != nil {
+		return err
+	}
+
+	if len(cols) > 0 {
+		tbl.SetColumns(cols)
+	}
+
+	return nil
+}
+
+// EnsureStructure ensures that the table identified by tableName contains the provided columns.
+//
+// createTemplates and addColumnTemplates are the templates which are executed in the event of table create or alter
+// (respectively).
+// metricsTableName and tagsTableName are passed to the templates.
+//
+// If the table cannot be modified, the returned column list is the columns which are missing from the table.
+func (tm *TableManager) EnsureStructure(
+	ctx context.Context,
+	db dbh,
+	tbl *tableState,
+	columns []utils.Column,
+	createTemplates []*template.Template,
+	addColumnsTemplates []*template.Template,
+	metricsTable *tableState,
+	tagsTable *tableState,
+) ([]utils.Column, error) {
+	// Sort so that:
+	//   * When we create/alter the table the columns are in a sane order (telegraf gives us the fields in random order)
+	//   * When we display errors about missing columns, the order is also sane, and consistent
+	utils.ColumnList(columns).Sort()
+
+	tbl.Lock()
+	tblColumns := tbl.Columns()
+	if tblColumns == nil {
+		// We don't know about the table. First try to query it.
+		if err := tm.refreshTableStructure(ctx, db, tbl); err != nil {
+			tbl.Unlock()
+			return nil, fmt.Errorf("querying table structure: %w", err)
+		}
+		tblColumns = tbl.Columns()
+
+		if tblColumns == nil {
+			// Ok, table doesn't exist, now we can create it.
+			if err := tm.executeTemplates(ctx, db, createTemplates, tbl, columns, metricsTable, tagsTable); err != nil {
+				tbl.Unlock()
+				return nil, fmt.Errorf("creating table: %w", err)
+			}
+
+			tblColumns = tbl.Columns()
+		}
+	}
+	tbl.Unlock()
```

Comment:
Does it make sense to move this part to an own function in order to be able to use
```golang
tbl.Lock()
defer tbl.Unlock()
```
making the locking more robust?

Reply:
If you prefer, I can do that. I think it's easier to follow the current function as is, as the locking earlier in the function is strongly related to the locking later in the function, and keeping it in one function makes it easier to see the whole picture. But that could be mitigated in multi-function by adding more comments.